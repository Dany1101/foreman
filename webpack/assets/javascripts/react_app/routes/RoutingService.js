import { visit } from '../../foreman_navigation';

let currentPath = window.location.href;

/**
 * a Helper function for rendering a route
 * @param {Function} renderFn - a component's rendering function
 * @param {Object} props - routing props
 * @param {Function} beforeRenderingCallback - a callback to run before the rendering, if it returns false the rendering will terminate
 */
export const renderRoute = (
  renderFn,
  props,
  beforeRenderingCallback = () => true
) => {
  const {
    location,
    location: { pathname, search },
  } = props;
  if (!beforeRenderingCallback(location)) return null;
  removeRailsContent();
  location && updatePath(`${pathname}${search}`);
  return renderFn(props);
};

export const fallbackRoute = () => {
  const nextPath = window.location.href;
  if (currentPath !== nextPath) {
    updatePath(nextPath);
    return visit(nextPath);
  }
  return null;
};

const updatePath = newPath => {
  if (newPath) currentPath = newPath;
};

const removeRailsContent = () => {
  const railsContainer = document.getElementById('rails-app-content');
  if (railsContainer) railsContainer.remove();
};
