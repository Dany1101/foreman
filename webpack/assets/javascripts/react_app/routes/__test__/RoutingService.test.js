import React from 'react';
import { renderRoute } from '../RoutingService';

jest.unmock('../RoutingService');

const props = { location: { pathname: '/test' } };
const renderFn = () => <div> Test </div>;
const callback = location => location.pathname === '/test';
const falseCallback = () => false;

describe('Routing Service', () => {
  it('rendering a route without a callback', () => {
    expect(renderRoute(renderFn, props)).toMatchSnapshot();
  });
  it('rendering a route with a callback', () => {
    expect(renderRoute(renderFn, props, callback)).toMatchSnapshot();
  });
  it('rendering a route with a falsy callback', () => {
    expect(renderRoute(renderFn, props, falseCallback)).toMatchSnapshot();
  });
});
