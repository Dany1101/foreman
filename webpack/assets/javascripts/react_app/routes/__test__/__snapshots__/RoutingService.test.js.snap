// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routing Service rendering a route with a callback 1`] = `
<div>
   Test 
</div>
`;

exports[`Routing Service rendering a route with a falsy callback 1`] = `null`;

exports[`Routing Service rendering a route without a callback 1`] = `
<div>
   Test 
</div>
`;
